File: packages/python/src/mcpsignals/intent_capture.py
"""Intent capture: optionally injects session_id/agent_id/intent into the
JSON schema a tool advertises, and strips those same keys back out of the
arguments before the user's tool handler runs, so the handler receives
exactly the arguments it would have received without the library.

Off by default. See prompt spec / README for the documented tradeoff: this
adds tokens to every advertised tool schema and asks the calling model to do
extra work it may ignore or fabricate.
"""

from collections.abc import Mapping
from typing import Any

INJECTED_KEYS = ("session_id", "agent_id", "intent")

# Length caps on the caller-controlled strings that land on an event: the
# three intent-capture values the calling agent supplies here, plus
# `tool_name` (read off the `tools/call` request) and `client_name`/
# `client_version` (declared in the `initialize` handshake), which
# `instrument.py` bounds with the same `bounded()`. Without them a sink that
# batches writes in a transaction (Cloudflare D1's `batch()`, for one) loses
# every unrelated event in the flush when one oversized value fails its
# statement. Capping before the event is built rather than in each sink means
# every sink, including third-party and future ones, inherits the bound.
#
# `intent` is prose and takes the same 2000 as `error_message`. The rest are
# identifiers - a UUID is 36 chars - so they take a much tighter cap. Oversized
# values are truncated rather than dropped: losing the tail of an id is cheaper
# than losing the event.
MAX_INTENT_LENGTH = 2000
MAX_IDENTIFIER_LENGTH = 128

_MAX_LENGTHS = {
    "session_id": MAX_IDENTIFIER_LENGTH,
    "agent_id": MAX_IDENTIFIER_LENGTH,
    "intent": MAX_INTENT_LENGTH,
}

_INJECTED_PROPERTIES = {
    "session_id": {
        "type": "string",
        "description": "Opaque id grouping this call with other calls in the same task, if known.",
    },
    "agent_id": {
        "type": "string",
        "description": "Opaque id identifying which agent is making this call, if multiple agents share a session.",
    },
    "intent": {
        "type": "string",
        "description": "Why you are calling this tool.",
    },
}


def enabled_for(
    tool_name: str, *, global_enabled: bool, overrides: Mapping[str, bool] | None
) -> bool:
    if overrides is not None and tool_name in overrides:
        return overrides[tool_name]
    return global_enabled


def inject_schema(input_schema: dict[str, Any] | None) -> dict[str, Any]:
    """Return a new JSON schema with the injected properties added. Never
    mutates the input schema in place.
    """
    schema = dict(input_schema) if input_schema else {"type": "object"}
    properties = dict(schema.get("properties") or {})
    for key, spec in _INJECTED_PROPERTIES.items():
        properties.setdefault(key, spec)
    schema["properties"] = properties
    return schema


def bounded(value: Any, max_length: int) -> str | None:
    """Return `value` cut to `max_length` characters, or None for a non-string.

    A caller can send any JSON type for these keys. Anything that isn't a
    string can't be length-bounded, so it's dropped rather than forwarded
    into an event field typed `str | None`. Python slices by code point, so
    the cut never lands inside a character.
    """
    if not isinstance(value, str):
        return None
    return value[:max_length]


def strip_injected(
    arguments: Mapping[str, Any] | None,
) -> tuple[dict[str, Any], dict[str, str | None]]:
    """Split raw arguments into (clean_arguments, extracted) where
    `extracted` holds session_id/agent_id/intent (None if absent, non-string
    or length-bounded per `_MAX_LENGTHS`) and `clean_arguments` has those
    three keys removed - exactly what the handler would have received without
    the library.
    """
    args = dict(arguments) if arguments else {}
    extracted = {key: bounded(args.pop(key, None), _MAX_LENGTHS[key]) for key in INJECTED_KEYS}
    return args, extracted

File: packages/node/src/intent-capture.ts
import { z } from 'zod';
import { boundedString, MAX_IDENTIFIER_LENGTH, MAX_INTENT_LENGTH } from './bounded.js';

export const INTENT_CAPTURE_KEYS = ['session_id', 'agent_id', 'intent'] as const;
export type IntentCaptureKey = (typeof INTENT_CAPTURE_KEYS)[number];

export type IntentCaptureOption = boolean | { tools?: Record<string, boolean> };

export function isIntentCaptureEnabled(
  option: IntentCaptureOption | undefined,
  toolName: string
): boolean {
  if (!option) return false;
  if (option === true) return true;
  return option.tools?.[toolName] ?? false;
}

const INTENT_FIELDS = {
  session_id: z.string().optional().describe('Groups this call with others in the same task.'),
  agent_id: z
    .string()
    .optional()
    .describe('Distinguishes this agent from others running in parallel.'),
  intent: z.string().optional().describe('Why you are calling this tool.')
};

/**
 * Extends a tool's input schema with optional session_id/agent_id/intent
 * fields so the client can see and populate them. Handles a full Zod object
 * schema (`.extend()`), a raw shape record (plain object merge, the legacy
 * `registerTool` form), and no schema at all (a fresh object schema with
 * just the three injected fields). Returns the input unchanged for any other
 * Standard Schema (e.g. Valibot/ArkType) — injection is skipped rather than
 * risking a crash on a schema shape this library doesn't understand; the
 * caller should treat that as "intent capture unavailable for this tool."
 */
export function injectIntentSchema<T>(inputSchema: T): T {
  if (inputSchema === undefined) {
    return z.object(INTENT_FIELDS) as unknown as T;
  }
  if (inputSchema instanceof z.ZodObject) {
    return inputSchema.extend(INTENT_FIELDS) as unknown as T;
  }
  if (
    typeof inputSchema === 'object' &&
    inputSchema !== null &&
    Object.values(inputSchema as Record<string, unknown>).every(v => v instanceof z.ZodType)
  ) {
    return { ...(inputSchema as Record<string, unknown>), ...INTENT_FIELDS } as unknown as T;
  }
  return inputSchema;
}

/** True if `injectIntentSchema` actually added the injected fields for this input. */
export function schemaSupportsInjection(inputSchema: unknown): boolean {
  if (inputSchema === undefined) return true;
  if (inputSchema instanceof z.ZodObject) return true;
  return (
    typeof inputSchema === 'object' &&
    inputSchema !== null &&
    Object.values(inputSchema as Record<string, unknown>).every(v => v instanceof z.ZodType)
  );
}

// The caps and the truncation helper live in bounded.ts because `instrument()`
// applies the same identifier cap to `client_name`/`client_version`; they are
// re-exported here so existing importers keep working.
export { MAX_INTENT_LENGTH, MAX_IDENTIFIER_LENGTH } from './bounded.js';

export interface ExtractedIntent {
  session_id: string | null;
  agent_id: string | null;
  intent: string | null;
  clean: Record<string, unknown>;
}

/**
 * Reads the injected keys off parsed args and returns args with those keys
 * removed. The extracted values are length-bounded; `clean` is untouched
 * beyond the removal, so the handler still receives exactly what it would
 * have received without the library.
 */
export function extractAndStripIntent(args: Record<string, unknown>): ExtractedIntent {
  const clean = { ...args };
  const session_id = boundedString(clean.session_id, MAX_IDENTIFIER_LENGTH);
  const agent_id = boundedString(clean.agent_id, MAX_IDENTIFIER_LENGTH);
  const intent = boundedString(clean.intent, MAX_INTENT_LENGTH);
  delete clean.session_id;
  delete clean.agent_id;
  delete clean.intent;
  return { session_id, agent_id, intent, clean };
}
